### ISSUE TITLE
SS4: Cannot sort `many_many` result when using `through`

### ISSUE THREAD

**Issue by chrispenny** [silverstripe/silverstripe-framework#7981]
Unsure if this is expected behaviour, or missing functionality. Please let me know :)

## Setup

I have a class (`Locale`) that has a `many_many` relationship to itself. This is being run `through` a class called `FallbackLocale`.

`Locale`:
```php
class Locale extends DataObject
{
    private static $has_many = [
        'FallbackLocales' => FallbackLocale::class . '.Parent',
    ];

    private static $many_many = [
        'Fallbacks' => [
            'through' => FallbackLocale::class,
            'from' => 'Parent',
            'to' => 'Locale',
        ],
    ];
}
```

`FallbackLocale`:
```php
class FallbackLocale extends DataObject
{
    private static $has_one = [
        'Parent' => Locale::class,
        'Locale' => Locale::class,
    ];

    private static $db = [
        'Sort' => 'Int',
    ];

    private static $default_sort = 'Sort';
}
```

## Scenario

I create a new `Locale` through a very standard Model Admin with a Gridfield.

I add two `Fallback()` records, and I sort them.

Whenever I use `$locale->Fallbacks()` (or `$locale->Fallback()->sort('Sort')`), the records are always in `ID` order, not the Gridfield `Sort` order.

## Workaround

If I use `$locale->FallbackLocales()` instead, I get the correct Gridfield sorting - however, I am then required to go an extra step to get the `Object` I'm actually interested in.

EG:
```php
foreach ($locale->FallbackLocales() as $throughObject) {
    $throughObject->Locale();
}
```

Or: I need to add a join before I sort.

## Pull requests

- [ ] #8417

**Comment by tractorcow** [comment]
Can you try

`$myLocale->FallbackLocales()->sort('"FallbackLocale"."Sort" ASC');`

Which, I'm afraid, might not be configurable via default_sort, but you can apply it to specific queries. :)

**Comment by mfendeksilverstripe** [comment]
@tractorcow It's actually `$myLocale->Fallbacks()->sort('"Fluent_FallbackLocale"."Sort" ASC');`

as that is the correct table name. Thanks for help Damian ;-)

**Comment by tractorcow** [comment]
so is it working now?

**Comment by mfendeksilverstripe** [comment]
Yes, it's working fine. We just need to remember to include proper table namespace to the `sort`

**Comment by tractorcow** [comment]
You know, I had a brain wave...

default_sort is set by DataQuery::initialiseQuery().

We could build in a bit of detection for many_many through, so that if the target default_sort is empty, it could check and inherit default_sort from the mapping table.


```php
if ($sort = singleton($this->dataClass)->config()->get('default_sort')) {
    $this->sort($sort);
} elseif ($sort = singleton($this->joinClass)->config()->get('default_sort')) {
    $this->sort($sort);
}
```

And have joinClass injected by many_many through list.

That would make your first example "just work".

**Comment by tractorcow** [comment]
Actually, we already inject the join table with ManyManyThroughQueryManipulator, maybe that same manipulator could add a sort (if none exists) as well?

**Comment by chrispenny** [comment]
I've been asked to add that: While we do have a workaround for this sorting issue (great for project code), Fluent (and maybe other modules) use `many_many_through`, and currently assume that this sorting functionality is working.

If possible, could we please make this fix upstream, rather than relying on the modules to apply the workaround?

CC: @frankmullenger

**Comment by robbieaverill** [comment]
We already have sorting tests using the table prefix as in Damian's suggestion above (`ManyManyThroughListTest::testSortingOnJoinTable()` as of #8417), and I've added another one into that PR which sorts without using a table alias as well.

As far as I can tell this functionality is working as expected.

The stub objects in the tests are the same as in the description of this issue, except that the parent/child objects aren't the same class (but I don't think that matters). I'll set up a new set of stubs that exactly match those in this example and see if it does.

**Comment by robbieaverill** [comment]
Ok I think I have found a bug. Here are my unit tests: https://gist.github.com/robbieaverill/1fedcc375da4703216a07dc37a2a8a6d

The first test fails, where you don't specify a sort in the ORM call. This _should_ default to using the sort order, but it doesn't look like it's doing that and is using the ID instead (as originally reported):

```
Starting test 'SilverStripe\ORM\Tests\MMTL7981Test::testDefaultSortOrder'.
F
Starting test 'SilverStripe\ORM\Tests\MMTL7981Test::testSortWithSpecificColumn'.
.
Starting test 'SilverStripe\ORM\Tests\MMTL7981Test::testSortWithSpecificColumnAscendingOrder'.
.
Starting test 'SilverStripe\ORM\Tests\MMTL7981Test::testWithSpecificSortColumnDescendingOrder'.
.                    
```

**Comment by robbieaverill** [comment]
With HasManyList the DataQuery looks up the default sort field during construction:

https://github.com/silverstripe/silverstripe-framework/blob/4/src/ORM/DataQuery.php#L174-L176

With ManyManyThroughList the default_sort could come from the joined class, not the data class.

I'll add some logic to `ManyManyThroughQueryManipulator` to handle this

**Comment by robbieaverill** [comment]
Ok I've adjusted the tests in #8417 so that they actually show the bug then pushed a fix for it to that PR

**Comment by chillu** [comment]
@micmania1 Could you peer review this please?

**Comment by robbieaverill** [comment]
Fixed with #8417. @micmania1 if you find anything you don't like please ping me on the PR and I'll address it separately

**Comment by chrispenny** [comment]
Hi team,

This doesn't appear to be working for me. I've made sure that I have the changes from:
https://github.com/silverstripe/silverstripe-framework/pull/8417/files

Same example, but more detail.

`Locale`:
```php
class Locale extends DataObject
{
    private static $table_name = 'Fluent_Locale';

    /**
     * @config
     * @var array
     */
    private static $db = [
        'Title' => 'Varchar(100)',
        'Locale' => 'Varchar(10)',
        'URLSegment' => 'Varchar(100)',
        'IsGlobalDefault' => 'Boolean',
    ];

    private static $has_many = [
        'FallbackLocales' => FallbackLocale::class . '.Parent',
    ];

    private static $many_many = [
        'Fallbacks' => [
            'through' => FallbackLocale::class,
            'from' => 'Parent',
            'to' => 'Locale',
        ],
    ];

    private static $default_sort = '"Fluent_Locale"."Locale" ASC';
}
```

`FallbackLocale` (I added the default sort here, but removing it has the same results):
```php
class FallbackLocale extends DataObject
{
    private static $db = [
        'Sort' => 'Int',
    ];

    private static $has_one = [
        'Parent' => Locale::class,
        'Locale' => Locale::class,
    ];

    private static $table_name = 'Fluent_FallbackLocale';

    private static $default_sort = 'Sort';
}
```

I have three Locales:

* International
* Mexico
* Argentina

Mexico has been set up with the following Fallback Locales (in order):

* Argentina
* International

These are my DB records:

`Fluent_Locale`:

| ID  | Title | Locale |
| --- | --- | --- |
| 1  | International | en_NZ | 
| 2  | Mexico  | es_MX |
| 3  | Argentina  | es_AR |

`Fluent_FallbackLocale`:

| ID  | Sort | ParentID | LocaleID |
| --- | --- | --- | --- |
| 1  | 2 | 2 | 1 |
| 2  | 1  | 2 | 3 |

I have a dev task to test:

```php
class TestLocaleTask extends BuildTask
{
    private static $segment = 'TestLocaleTask';

    public function run($request): void
    {
        /** @var Locale $locale */
        $locale = Locale::get()->filter('Locale', 'es_MX')->first();

        Debug::message($locale->Fallbacks()->count());

        foreach ($locale->Fallbacks() as $fallback) {
            Debug::message($fallback->Locale);
        }
    }
}
```

The output of this is:
```
Debug (line 48 of TestLocaleTask.php): 2

Debug (line 51 of TestLocaleTask.php): es_AR

Debug (line 51 of TestLocaleTask.php): en_NZ
```

Appears to be working! But then I change Mexico's Fallback order to:

* International
* Argentina

`Fluent_FallbackLocale`:

| ID  | Sort | ParentID | LocaleID |
|  --- | --- | --- | --- |
| 1  | 1 | 2 | 1 |
| 2  | 2  | 2 | 3 |

I would not expect the order to switch.

I run the dev task again, and the output is still:
```
Debug (line 48 of TestLocaleTask.php): 2

Debug (line 51 of TestLocaleTask.php): es_AR

Debug (line 51 of TestLocaleTask.php): en_NZ
```

This affects Fluent heavily because the module uses `$locale->Fallbacks()` to determine it's "chain", which means you get incorrect content.
